src/Features/Core/Portable/ConvertIfToSwitch/AbstractConvertIfToSwitchCodeRefactoringProvider.Analyzer.cs

```diff
@@ -0,0 +1,397 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+#nullable enable
+
+using System;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Operations;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.ConvertIfToSwitch
+{
+    using static BinaryOperatorKind;
+
+    internal abstract partial class AbstractConvertIfToSwitchCodeRefactoringProvider<
+        TIfStatementSyntax, TExpressionSyntax, TIsExpressionSyntax, TPatternSyntax>
+    {
+        // Match the following pattern which can be safely converted to switch statement
+        //
+        //    <if-statement-sequence>
+        //        : if (<section-expr>) { <unreachable-end-point> }, <if-statement-sequence>
+        //        : if (<section-expr>) { <unreachable-end-point> }, ( return | throw )
+        //        | <if-statement>
+        //
+        //    <if-statement>
+        //        : if (<section-expr>) { _ } else <if-statement>
+        //        | if (<section-expr>) { _ } else { _ }
+        //        | if (<section-expr>) { _ }
+        //        | { <if-statement-sequence> }
+        //
+        //    <section-expr>
+        //        : <section-expr> || <pattern-expr>
+        //        | <pattern-expr>
+        //
+        //    <pattern-expr>
+        //        : <pattern-expr> && <expr>                         // C#
+        //        | <expr0> is <pattern>                             // C#
+        //        | <expr0> is <type>                                // C#
+        //        | <expr0> == <const-expr>                          // C#, VB
+        //        | <expr0> <comparison-op> <const>                  //     VB
+        //        | ( <expr0> >= <const> | <const> <= <expr0> )
+        //           && ( <expr0> <= <const> | <const> >= <expr0> )  //     VB
+        //        | ( <expr0> <= <const> | <const> >= <expr0> )
+        //           && ( <expr0> >= <const> | <const> <= <expr0> )  //     VB
+        //
+        internal abstract class Analyzer
+        {
+            private SyntaxNode? _switchTargetExpression;
+            private readonly ISyntaxFactsService _syntaxFacts;
+
+            protected Analyzer(ISyntaxFactsService syntaxFacts)
+            {
+                _syntaxFacts = syntaxFacts;
+            }
+
+            public (ImmutableArray<AnalyzedSwitchSection>, SyntaxNode) AnalyzeIfStatementSequence(ReadOnlySpan<IOperation> operations)
+            {
+                var sections = ArrayBuilder<AnalyzedSwitchSection>.GetInstance();
+                if (!ParseIfStatementSequence(operations, sections, out var defaultBodyOpt))
```

Comment:
consider renaming to DecomposeXXX

Reply:
ah, *parse* only makes sense for text =) maybe "match" would be more suitable since we're rather matching against some pattern than just "breaking down" things

Reply:
'match' would work for me :)  but it's a minor thing anyways.  if you like 'parse' that's fine :)